common/crypto/attestation-api/evidence/verify-evidence.cpp

```diff
@@ -0,0 +1,257 @@
+
+    return true;
+
+err:
+    return false;
+}
+
+bool extract_hex_from_report(
+    const std::string& ias_report, size_t offset, size_t size, std::string& hex)
+{
+    std::string b64quote;
+    ByteArray bin_quote;
+    sgx_report_body_t* rb;
+    ByteArray ba;
+
+    JsonValue root_value = json_parse_string(ias_report.c_str());
+    JSON_Object* root_object = json_value_get_object(root_value);
+    COND2LOGERR(root_object == NULL, "bad ias json report");
+
+    b64quote = json_object_get_string(root_object, "isvEnclaveQuoteBody");
+    COND2LOGERR(b64quote.length() == 0, "no isvEnclaveQuoteBody");
+
+    bin_quote = Base64EncodedStringToByteArray(b64quote);
+    COND2LOGERR(bin_quote.size() != offsetof(sgx_quote_t, signature_len), "unexpected quote size");
+    ba = ByteArray(bin_quote.data() + offset, bin_quote.data() + offset + size);
+    hex = ByteArrayToHexEncodedString(ba);
+
+    return true;
+
+err:
+    return false;
+}
+
+bool verify_ias_evidence(
+    ByteArray& evidence, ByteArray& expected_statement, ByteArray& expected_code_id)
+{
+    std::string evidence_str((char*)evidence.data(), evidence.size());
+    std::string expected_hex_id((char*)expected_code_id.data(), expected_code_id.size());
+
+    std::string ias_signature, ias_certificates, ias_report;
+    std::vector<std::string> ias_certificate_vector;
+
+    // get evidence data
+    COND2ERR(
+        false == unwrap_ias_evidence(evidence_str, ias_signature, ias_certificates, ias_report));
+
+    // split certs
+    COND2ERR(false == split_certificates(ias_certificates, ias_certificate_vector));
+
+    {
+        // verify report status
+        const int group_out_of_date_ok = 1;
+        COND2LOGERR(VERIFY_SUCCESS != verify_enclave_quote_status(ias_report.c_str(),
+                                          ias_report.length(), group_out_of_date_ok),
+            "invalid quote status");
+    }
+
+    {
+        // check root cert
+        const int root_certificate_index = 1;
+        COND2LOGERR(VERIFY_SUCCESS != verify_ias_certificate_chain(
+                                          ias_certificate_vector[root_certificate_index].c_str()),
+            "invalid root certificate");
+    }
+
+    {
+        // check signing cert
+        const int signing_certificate_index = 0;
+        COND2LOGERR(
+            VERIFY_SUCCESS != verify_ias_certificate_chain(
+                                  ias_certificate_vector[signing_certificate_index].c_str()),
+            "invalid signing certificate");
+
+        // check signature
+        COND2LOGERR(VERIFY_SUCCESS != verify_ias_report_signature(
+                                          ias_certificate_vector[signing_certificate_index].c_str(),
+                                          ias_report.c_str(), ias_report.length(),
+                                          (char*)ias_signature.c_str(), ias_signature.length()),
+            "invalid report signature");
+    }
```

Comment:
> though IAS has an intermediary CA but i guess it doesn't?

Right, it does not. It has the root CA certificate and the signing one.

> the first verify_ias_certificate_chain effectively checks identity of the built-in root-key and this key in a bit more expensive way?

The received certificate and the hard-coded one should be the same self-signed certificate.
The `verify_ias_certificate_chain` here checks if the hard-coded one certifies the received one -- so the public key of the hard-coded one verifies the signature in the latter.
Maybe you could do a string comparison, but it's brittle.
I don't think you can meaningfully check the received one alone because it's self-signed (i.e., you have no trust anchor).

Does this answer?

Reply:
yep, answers. Strictly speaking you wouldn't really have to test root (which confused me a bit) but it certainly doesn't hurt and maybe makes a root CA mismatch (vs, say cert corruption or cert mixups) more explicit